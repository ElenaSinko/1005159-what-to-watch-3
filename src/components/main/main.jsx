import React, {PureComponent} from "react";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";
import FilmsList from "../films-list/films-list.jsx";
import {ServerIsNotAvailable} from "../server-is-not-available/server-is-not-available.jsx";
import GenresList from "../genres-list/genres-list.jsx";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer/application-state/application-state.js";
import {AuthorizationStatus} from "../../reducer/user/user.js";
import {getAuthorizationStatus, getUserIMG} from "../../reducer/user/selectors.js";
import {Operation as UserOperation} from "../../reducer/user/user.js";
import {getFilmCards, getGenre, getFilmsToShow, getServerAvailability, getPromoFilm} from "../../reducer/application-state/selectors.js";
import {unique} from "../../utils.js";
import {Button} from "../button/button.jsx";
import VideoPlayerFullScreen from "../video-player-full-screen/video-player-full-screen.jsx";

class Main extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      playerIsWorking: false,
    };
  }

  render() {
    const {filmCards, onGenreTitleClick, showMore, filmsToShow, authorizationStatus, serverIsAvailable, promoFilm, userIMG} = this.props;
    if (!serverIsAvailable) {
      return <ServerIsNotAvailable />;
    }
    if (typeof filmCards === `undefined`) {
      return <div>Loading...</div>;
    }
    const currentCards = filmCards.slice(0, filmsToShow);
    const genres = [`All genres`].concat(unique(filmCards.map((movieCard) => movieCard.genre)));
    return <React.Fragment>
      {!this.state.playerIsWorking &&
        <React.Fragment>
          <section className="movie-card">
            <div className="movie-card__bg">
              <img src={promoFilm.movieBG} alt={promoFilm.name}/>
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <header className="page-header movie-card__head">
              <div className="logo">
                <a className="logo__link">
                  <span className="logo__letter logo__letter--1">W</span>
                  <span className="logo__letter logo__letter--2">T</span>
                  <span className="logo__letter logo__letter--3">W</span>
                </a>
              </div>
              {authorizationStatus === AuthorizationStatus.AUTH && <div className="user-block">
                <div className="user-block__avatar">
                  <img src={`https://htmlacademy-react-3.appspot.com/` + userIMG} alt="User avatar" width="63" height="63"/>
                </div>
              </div>}
              {authorizationStatus === AuthorizationStatus.NO_AUTH && <div className="user-block">
                <Link to={`/sign-in`} style={{textDecoration: `none`}}>
                  <div href="sign-in.html" className="user-block__link">Sign in</div>
                </Link>
              </div>}
            </header>
            <div className="movie-card__wrap">
              <div className="movie-card__info">
                <div className="movie-card__poster">
                  <img src={promoFilm.img} alt="The Grand Budapest Hotel poster" width="218" height="327"/>
                </div>
                <div className="movie-card__desc">
                  <h2 className="movie-card__title">{promoFilm.name}</h2>
                  <p className="movie-card__meta">
                    <span className="movie-card__genre">{promoFilm.genre}</span>
                    <span className="movie-card__year">{promoFilm.movieYear}</span>
                  </p>

                  <div className="movie-card__buttons">
                    <button onClick={() => {
                      this.setState({playerIsWorking: true});
                    }} className="btn btn--play movie-card__button" type="button">
                      <svg viewBox="0 0 19 19" width="19" height="19">
                        <use xlinkHref="#play-s"></use>
                      </svg>
                      <span>Play</span>
                    </button>
                    <button className="btn btn--list movie-card__button" type="button">
                      <svg viewBox="0 0 19 20" width="19" height="20">
                        <use xlinkHref="#add"></use>
                      </svg>
                      <span>My list</span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </section>

          <div className="page-content">
            <section className="catalog">
              <h2 className="catalog__title visually-hidden">Catalog</h2>

              <GenresList genres={genres} onGenreTitleClick={onGenreTitleClick}/>
              <FilmsList smallMovieCards={currentCards} />
              {filmsToShow < filmCards.length && <Button onClick={showMore}/>}
            </section>

            <footer className="page-footer">
              <div className="logo">
                <a className="logo__link logo__link--light">
                  <span className="logo__letter logo__letter--1">W</span>
                  <span className="logo__letter logo__letter--2">T</span>
                  <span className="logo__letter logo__letter--3">W</span>
                </a>
              </div>

              <div className="copyright">
                <p>Â© 2019 What to watch Ltd.</p>
              </div>
            </footer>
          </div></React.Fragment>}
      {this.state.playerIsWorking && <VideoPlayerFullScreen playerIsWorking={this.state.playerIsWorking} src={filmCards[0].srcFullVideo} name={filmCards[0].name} poster={filmCards[0].img} closeVideoPlayerFullScreen={() => {
        this.setState({playerIsWorking: false});
      }}/>}
    </React.Fragment>;
  }
}

Main.propTypes = {
  authorizationStatus: PropTypes.string.isRequired,
  login: PropTypes.func.isRequired,
  filmCards: PropTypes.array,
  onGenreTitleClick: PropTypes.func,
  showMore: PropTypes.func,
  filmsToShow: PropTypes.number,
  onPlayButtonClick: PropTypes.func,
  serverIsAvailable: PropTypes.bool,
  promoFilm: PropTypes.object,
  userIMG: PropTypes.string,
};

const mapStateToProps = (state) => ({
  filmCards: getFilmCards(state),
  filmsToShow: getFilmsToShow(state),
  genre: getGenre(state),
  authorizationStatus: getAuthorizationStatus(state),
  serverIsAvailable: getServerAvailability(state),
  promoFilm: getPromoFilm(state),
  userIMG: getUserIMG(state),
});

const mapDispatchToProps = (dispatch) => ({
  onGenreTitleClick(genre) {
    dispatch(ActionCreator.changeGenre(genre));
  },
  showMore() {
    dispatch(ActionCreator.showMoreFilms());
  },
  login(authData) {
    dispatch(UserOperation.login(authData));
  },
});

const connectedComponent = connect(mapStateToProps, mapDispatchToProps)(Main);
export {connectedComponent as Main};

