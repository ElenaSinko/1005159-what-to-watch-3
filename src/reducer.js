import movieCards from "./mocks/films.js";
import {extend} from "./utils.js";

const initialState = {
  genre: `All genres`,
  filmCards: movieCards.slice(0, 8),
};

const ActionType = {
  CHANGE_GENRE: `CHANGE_GENRE`,
  SHOW_MORE_FILMS: `SHOW_MORE_FILMS`,
};

const ActionCreator = {
  changeGenre: (genre) => ({
    type: ActionType.CHANGE_GENRE,
    genre,
  }),
  showMoreFilms: () => ({
    type: ActionType.SHOW_MORE_FILMS
  }),
};


const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_GENRE:
      if (action.genre === `All genres`) {
        return initialState;
      }
      return extend(state, {
        genre: action.genre,
        filmCards: movieCards.filter((it) => it.genre === action.genre).slice(0, 8),
      });
    case ActionType.SHOW_MORE_FILMS:
      return initialState;
    default:
      return state;
  }
};

export {reducer, ActionType, ActionCreator};
