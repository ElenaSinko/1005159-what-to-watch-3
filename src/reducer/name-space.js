const NameSpace = {
  APPLICATION_STATE: `APPLICATION_STATE`,
  USER: `USER`,
};

export default NameSpace;
