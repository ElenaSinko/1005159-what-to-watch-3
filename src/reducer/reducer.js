import {combineReducers} from "redux";
import {reducer as applicationState} from "./application-state/application-state.js";
import NameSpace from "./name-space.js";


export default combineReducers({
  [NameSpace.APPLICATION_STATE]: applicationState,
});
