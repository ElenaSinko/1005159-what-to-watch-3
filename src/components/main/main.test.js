import React from "react";
import renderer from "react-test-renderer";
import {Main} from "./main.jsx";
import {BrowserRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {applyMiddleware, createStore} from "redux";
import reducer from "./../../reducer/reducer.js";
import thunk from "redux-thunk";
import {createAPI} from "../../api";


const api = createAPI(() => {});

const store = createStore(
    reducer,
    applyMiddleware(thunk.withExtraArgument(api))
);

const smallMovieCards = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    imgPrev: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    id: 1,
    srcFullVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Drama`,
    movieYear: 2014,
    duration: 137,
    moviePoster: `/img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    movieBG: `/img/bg-the-grand-budapest-hotel.jpg`,
    rating: `8,9`,
    movieRatingLevel: `Very good`,
    movieRatingCount: `240 ratings`,
    movieDescription: ``,
    director: `Wes Andreson`,
    description: `Wes Andreson`,
    starring: `Bill Murray, Edward Norton, Jude Law, Willem Dafoe and other`,
    onOverviewTabClick: () => {},
  }, {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    imgPrev: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    id: 1,
    srcFullVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Drama`,
    movieYear: 2014,
    duration: 137,
    moviePoster: `/img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    movieBG: `/img/bg-the-grand-budapest-hotel.jpg`,
    rating: `8,9`,
    movieRatingLevel: `Very good`,
    movieRatingCount: `240 ratings`,
    movieDescription: ``,
    director: `Wes Andreson`,
    description: `Wes Andreson`,
    starring: `Bill Murray, Edward Norton, Jude Law, Willem Dafoe and other`,
    onOverviewTabClick: () => {},
  }
];

it(`Render main screen`, () => {
  const tree = renderer
    .create(<Provider store={store}>
      <BrowserRouter>
        <Main
          filmCards={smallMovieCards} onGenreTitleClick={() => {}}
        />
      </BrowserRouter>
    </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
