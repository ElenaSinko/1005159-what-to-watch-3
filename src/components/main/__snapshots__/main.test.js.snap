// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render main screen 1`] = `
<div>
  Loading...
</div>
`;
